File: src/main/scala/com/landoop/data/generator/stocks/Tick.scala
package com.landoop.data.generator.stocks

import com.univocity.parsers.csv.{CsvParser, CsvParserSettings}

import scala.collection.JavaConverters._

case class Exchange(name: String, symbol: String)

object Exchange {
  def forSymbol(symbol: String): Exchange = {
    val name = symbol match {
      case "A" => "NYSE MKT"
      case "N" => "New York Stock Exchange (NYSE)"
      case "P" => "NYSE ARCA"
      case "Z" => "BATS Global Markets (BATS)"
      case "V" => "Investors' Exchange, LLC (IEXG)"
    }
    Exchange(name, symbol)
  }
}

case class Stock(symbol: String, name: String, etf: Boolean, exchange: Exchange, lotSize: Int)

object Stock {

  def load: Seq[Stock] = {

    val settings: CsvParserSettings = new CsvParserSettings
    settings.getFormat.setLineSeparator("\n")
    settings.getFormat.setDelimiter('|')

    val parser: CsvParser = new CsvParser(settings)
    parser.parseAllRecords(getClass.getResourceAsStream("/otherlisted.txt")).asScala
      .map { record =>
        Stock(
          record.getString("NASDAQ Symbol"),
          record.getString("Security Name"),
          record.getBoolean("ETF"),
          Exchange.forSymbol(record.getString("Exchange")),
          record.getInt("Round Lot Size")
        )
      }
  }
}

case class Tick(symbol: String, name: String, category: String, big: Double, ask: Double, etf: Boolean, lotSize: Int, exchange: Exchange)

File: datagen-redis/src/main/kotlin/io/lenses/datagen/redis/Tick.kt
package io.lenses.datagen.redis

import com.univocity.parsers.csv.CsvParser
import com.univocity.parsers.csv.CsvParserSettings
import java.util.*

data class Stock(val symbol: String, val name: String, val etf: Boolean, val exchange: Exchange, val lotSize: Int) {
  companion object {

    val stocks: Array<Stock> = CsvParserSettings().run {
      format.setLineSeparator("\n")
      format.delimiter = '|'
      isHeaderExtractionEnabled = true

      val parser = CsvParser(this)
      parser.parseAllRecords(this::class.java.getResourceAsStream("/otherlisted.txt")).map {
        Stock(
            it.getString("NASDAQ Symbol"),
            it.getString("Security Name"),
            it.getString("ETF") == "Y",
            Exchange.forSymbol(it.getString("Exchange")),
            it.getInt("Round Lot Size")
        )
      }.toTypedArray()
    }
  }
}

data class Exchange(val name: String, val symbol: String) {
  companion object {
    fun forSymbol(symbol: String): Exchange {
      val name = when (symbol) {
        "A" -> "NYSE MKT"
        "N" -> "New York Stock Exchange (NYSE)"
        "P" -> "NYSE ARCA"
        "Z" -> "BATS Global Markets (BATS)"
        "V" -> "Investors' Exchange, LLC (IEXG)"
        else -> "Nasdaq"
      }
      return Exchange(name, symbol)
    }
  }
}

data class Tick(val symbol: String, val name: String, val category: String, val bid: Double, val ask: Double, val etf: Boolean, val lotSize: Int, val exchange: Exchange) {

  fun toMap(): Map<String, String> {
    return mapOf(
        "symbol" to symbol,
        "name" to name,
        "category" to category,
        "bid" to bid.toString(),
        "ask" to ask.toString(),
        "etf" to etf.toString(),
        "lotSize" to lotSize.toString(),
        "exchange.name" to exchange.name,
        "exchange.symbol" to exchange.symbol
    )
  }

  companion object {
    private val random = Random()
    fun gen(): Tick {
      val stock = Stock.stocks[random.nextInt(Stock.stocks.size)]
      val price = random.nextDouble() * 1000
      val spread = price / random.nextDouble()
      val bid = price - spread
      val ask = price + spread
      return Tick(
          stock.symbol,
          stock.name,
          listOf("N", "Q").shuffled().first(),
          bid,
          ask,
          stock.etf,
          random.nextInt(100) * 100,
          stock.exchange
      )
    }
  }
}